stdlib/src/collections/string.mojo

```diff
@@ -1542,7 +1543,9 @@ struct String(
         )
 
     @always_inline
-    fn as_string_slice(ref [_]self) -> StringSlice[__origin_of(self)]:
+    fn as_string_slice(
+        ref [_]self,
+    ) -> StringSlice[_lit_mut_cast[__origin_of(self), False].result]:
```

Comment:
Because otherwise we'd need to add the equivalent of `Span.get_immutable()` and do `string_var.as_string_slice().get_immutable()` which is way too verbose. And I figured that to keep the scope f this PR small it might be better to just force the API to return immutable slices. I'm fine with doing it either way, WDYT?

Reply:
Reverted the changes using rebind in e51154c59c222b0c8abfb67afe2b1ed8fbf26df9